test/integration/channels.js

```diff
@@ -2975,3 +2975,35 @@ describe('Channel - isUpToDate', async () => {
 		).to.be.equal(-1);
 	});
 });
+
+describe('Ensure single channel per cid on client', async () => {
+	it('channel created using id', async () => {
+		const userIdVish = 'vishal';
+		const userIdAmin = 'amin';
+		await createUsers([userIdVish, userIdAmin]);
+
+		const clientVish = await getTestClientForUser(userIdVish);
+		const channelVishId = uuidv4();
+		const channelVish_copy1 = clientVish.channel('messaging', channelVishId);
+		await channelVish_copy1.watch();
+
+		const channelVish_copy2 = clientVish.channel('messaging', channelVishId);
+
+		expect(channelVish_copy1).to.be.equal(channelVish_copy2);
+	});
+	it('channel created using member list', async () => {
+		const userIdVish = 'vishal';
+		const userIdAmin = 'amin';
+		await createUsers([userIdVish, userIdAmin]);
+
+		const clientVish = await getTestClientForUser(userIdVish);
+		const channelVish_copy1 = clientVish.channel('messaging', {
+			members: ['amin', 'vishal'],
+		});
+		await channelVish_copy1.watch();
+		const channelVish_copy2 = clientVish.channel('messaging', {
+			members: ['vishal', 'amin'],
+		});
```

Comment:
kinda an edge case to this fix which itself is an edge case but if you call `channel.watch` or `channel.create` after instantiating the second channel you still end up with a duplicate channel in the state. (this can actually happen if different parts of the app instantiate channels at the same time)

```js
const channelVish_copy1 = clientVish.channel('messaging', { members: ['amin', 'vishal'] });
const channelVish_copy2 = clientVish.channel('messaging', { members: ['amin', 'vishal'] });
// channelVish_copy1 !== channelVish_copy2

await channelVish_copy1.watch();
await channelVish_copy2.watch();

// channelVish_copy1 !== channelVish_copy2

```

Reply:
hmm probably we should early return from `watch()` function if its already being watched
```
watch() {
  if (this.initialized) {
    return;
  }

  ... // rest of the logic
}
```

Reply:
yea it's a potential fix but there could be changes in the option param which cause other issues. To cover this we need to make some changes to the new code you implemented to also check for channels that don't have their `cid` yet but have the same member list

Reply:
Makes sense!! Have updated the PR.

So when channel doesn't have `cid` yet, I am storing it in `activeChannels` as following:

```js
{
  "messaging:amin,vishal": uniqueChannelAminVishal,
  "messaging:k4j2n3k4j2n3kn42k3": channelWithId
}
```

When cid gets set on channel, I replace `messaging:amin,vishal` with actual cid generated from backend.

```js
{
  "messaging:!members:kajshdkjsakjsdkjas": uniqueChannelAminVishal,
  "messaging:k4j2n3k4j2n3kn42k3": channelWithId
}
```

Also updated tests to reflect this case :)